### PR TITLE
Docker

### LINKED ISSUE #1: Docker
- [x] Create a `Dockerfile`.
- [x] Install all the dependencies and run the server in production.
- [x] Make sure you can build the Docker image on your local machine.
- [x] Run a container from that Docker image on your local machine, and expose the relevant port so you send a request to the root endpoint to make sure you get `{hello: "world"}` in response.

---

Initial instructions:
1. `clone` the repo to you local machine.
2. Install everything and make sure you can run the project (call me if you have any questions).
3. Install a REST/HTTP client like [insomnia](https://insomnia.rest) to send a request to the root endpoint and make sure you get `{hello: "world"}` in response (a screenshoot from my insomnia is below).
4. Create a new branch and move to it, the name of the branch should be this issue number (`git checkout -b 1`).
5. Create the `Dockerfile` file and add it to the staging area (`git add .`).
6. Commit the changes (`git commit -m "Create a Dockerfile file"`).
7. Push the new branch (`git push origin 1`).
8. Go to the repo GitHub page on your browser and create the PR (Pull request) (call me if you need any help with this).
9. Continue work on the other tasks of that issue (you can mark the first task as done). 

---

Things to pay attention to:
- When I write in the issue to "make sure" I mean you need to check it (for example: that something is working), but I also want to see it with my own eyes. I'm lazy as fuck, so when you make sure that something is working, screenshot it (your terminal for example or GCP UI, etc) and upload that screenshot to the PR (Pull Request).
- My insomnia:
Default_–_book_store

### PR DESCRIPTION
**Pull request by shahargalon**
Hi,
I changed the python version on the Pipfile and run this on my machine and its work 

image

now I am tring to install everything on the official python docker and run it locally (right now i am experience some issues)
and when I will manage to do this I will right the step on the Dockerfile

### CODE REVIEW

**Comment by nirgn975** [comment]
@shahargalon 
1. Change the PR name to be the same as the issue name.
2. Put me on "Reviewers", because otherwise I don't get any notification about this PR (you can also mention me with `@` in the message, like I did above).

**Comment by shahargalon** [comment]
Hi, I generate the requirements.txt and wrote the Dockerfile (that lays on the official python Docker).
the image is 920MB. and its run perffect
image

**Comment by nirgn975** [comment]
Looking good.

1. in future PRs (Pull Requests), there is a _"Linked issues"_ at the right menu (almost at the bottom of the menu), choose the issue that this PR is related to. This will create a link between the PR and the Issue (in the issue page and in the PR page, so we can go between them without searching). Also, it'll close the issue automatically once we'll close the PR (when we'll `merge` it). No need to do this now, I did it this time.
2. Mark the tasks you finished in the issue as done.
3. After reading a bit, we have 2 options:
  3.1 We need to create the CI script (GitHub action workflow) so every time someone does a PR (Pull Request) with a change to the `Pipfile` it'll auto generate a new `requirements.txt` file.
  3.2 I read that there is a way to use `pipenv` and tells it to install the dependencies (packages our project depend on, the ones in the `Pipfile`) right in the system (and not in the virtual environment). This way, we don't need the `requirements.txt` file anymore, we only need to install `pipenv` in the Docker image (don't `shell` into the virtual environment), and then to install the dependencies (look at this: https://stackoverflow.com/a/49705601/2040160).

So, which way to do want to go related to 3 ? it's your show, choose.

**Comment by shahargalon** [comment]
I wrote another Docker file who use the pipenv and the Pipfile (the size is little bit more then the fisrt one 965MB because the pipenv).
do you want me do remove the requirements.txt file? and also replace the Dockerfile to the new one on the same brunch?

**Comment by nirgn975** [comment]
@shahargalon Yeah. I mean, if you did that, we don't need the `requirements.txt` anymore.. And we'll use the new `Dockerfile` - no reason to use the one with the `requirements.txt`..

This branch and PR, should solve all the tasks in the issue, I don't care how, the implementation details are your problem (:

When you finish all the things, and the PR is ready to be merged, tag me and write that this PR is ready to review. I'll review it, and if everything is good I'll merge it to `master` and open a new issue (if there is something to fix, I'll write it here, when I review the code).

**Comment by shahargalon** [comment]
Hi @nirgn975 ,
I finished the Dockerfile, the PR should ready to be merged 
Please review it and let me know if all good

thanks

File: Dockerfile
Comment (nirgn975) [inline Dockerfile]: Delete this whitespace (empty line).
You don't need 2 empty lines, one is enough, unless there is Docker convention I don't know of that tells to keep 2 empty lines in a case like this.
If there is such a convention, add a comment here and point me to it (paste a link or something).

File: Dockerfile
Comment (nirgn975) [inline Dockerfile]: Delete this whitespace.

File: Dockerfile
Comment (nirgn975) [inline Dockerfile]: Delete this whitespace.

**Review by nirgn975 — CHANGES_REQUESTED** [review]
I Deliberately pushed a commit right to the `master` that has some changes with the files you need, so you need to "Resolve conflicts" (see a button in the bottom of the PR to do that on GitHub UI).

This is like a team member working on the project and his PR merged between the time you `checkout` to a new branch and the time you want to `merge` yours.

Also, you can see above that I review the code and asked for 3 changes.
And, notice I changed things in the code, you need to adjust you Dockerfile now..

File: Pipfile
Comment (shahargalon) [inline Pipfile]: @nirgn975 what is this package???,
when I create the Pipfile.lock to the new Pipfile I get an err msg: AttributeError: 'NoneType' object has no attribute 'url'

when I hashed this package it work fine. I couldn't manage to see whats wrong

**Comment by shahargalon** [comment]
Hi @nirgn975 
can you take a look on the Pipfile I added a note in this file on some issue that I cant understand..

File: Dockerfile
Comment (nirgn975) [inline Dockerfile]: I didn't mean that you remove all the spaces, just add one and not two together.
Also, you can group commands together if you think they should be together (related or kinda the same), and use one whitespace (`enter`) to separate some of them (but not two whitespaces).

File: Dockerfile
Comment (nirgn975) [inline Dockerfile]: The `--ignore-pipfile` means it's installing the dependencies from the `Pipfile.lock` file, so you don't need the `Pipfile` in the docker.

File: Dockerfile
Comment (nirgn975) [inline Dockerfile]: You need to add the `src` directory.

File: Dockerfile
Comment (nirgn975) [inline Dockerfile]: You need to fix this command, this won't work now..

**Review by nirgn975 — CHANGES_REQUESTED** [review]

File: Pipfile
Comment (nirgn975) [reply Pipfile]: 1. We write here in [markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet), so you can use the "```" (three backticks) to start a code block and end it with another three of them.
2. Why do you create a new `Pipfile.lock`? You don't need to create the file, we already have it, it's generated automatcally every time someone do install/uninstall with pipenv.
3. You don't "hashed" it, you "commented" it (the hash here is to write comments, so you turn this line into a comment and that's why the system is ignoring it now).

File: Dockerfile
Comment (shahargalon) [reply Dockerfile]: If I copy only the Pipfile.lock without the Pipfile, the "pipenv install --system  --ignore-pipfile" does not work.
anyway I copied all project directory to the docker image, because I still need to figure out how to copy a specific directors and keep the structure formation.

Comment (nirgn975) [reply Dockerfile]: Actually do that, copy it all, but use the `.dockerignore` file (Google it..)

Comment (shahargalon) [reply Dockerfile]: OK WOW! this is cool feature `.dockerignore` did it..
can I marge this branch to the master and move to the next issue?
on the next issue I need to open a new branch??

Comment (nirgn975) [reply Dockerfile]: If you think the PR is ready to be reviewed write me a comment that says it's ready..
I'll review it now..

File: Dockerfile
Comment (nirgn975) [inline Dockerfile]: Delete this whitespace, you have two whitespace lines (9 and 10), you only need one at the EOF (**E**nd **O**f **F**ile)

**Review by nirgn975 — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (shahargalon) [reply Dockerfile]: done,
the PR is ready to be merged

**Review by nirgn975 — APPROVED** [review]
Great. You can `merge` it.
Click on the little arrow in the green button and then choose _"Squash and merge"_. and then click it.

**Pull request merged**
